### PR TITLE
Ili9341 8bit indexed mode pt1

### PR DESCRIPTION
**Pull request by davet2001**
# What does this implement/fix? 
Adds 8bit palette mode to ili9341 displays for better color quality.

As discussed in [#1435](https://github.com/esphome/feature-requests/issues/1451), the color support on in esphome could be improved.  The display supports 16bit (5,6,5 RGB) color, but the display buffer is only 8bits per pixel due to the ESP32 limited RAM.  This means that every pixel on the screen is first converted to 8 bit, stored in a buffer, then later converted to 16bits for display.  Obviously this loses quality.  Example image before this PR:

 
This PR adds a palette mode to the display, where the palette can be defined in the yaml file.  To keep the PR short the only palette modes supported are `NONE` (the default) and `GRAYSCALE`.  Example image in with grayscale palette:

In part 2 of the PR I will add an `IMAGE` palette mode, where the 265 color palette is generated in python using a user supplied image file and saved as a const in flash in the ESP.  If possible also an 'AUTO' mode where it bases the palette on all images defined in the yaml file.

This feature is implemented on the ili9341 display only, but if it works well, it could be made more generic and applied to other displays.

The overall aim is to make it really simple: "Just add `color_palette: AUTO` to make the display look better", without needing to really understand what's going on underneath.

Following this PR, further improvements could be done:
 - store the palette as 16bit values to avoid the need to convert to Color type first (speed improvement, although even without optimising this way it seems very fast).
 - store images as 8bit indexed data rather than RGB (reduces flash size for images by 66%, and possibly speed improvement also).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** related issue https://github.com/esphome/feature-requests/issues/1451

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2087

## Test Environment

- [x] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
image:
  - file: "sample_100x100.png"
    id: myimage
    resize: 100x100
    type: RGB24

display:
  - platform: ili9341
    model: TFT 2.4
    cs_pin: 5
    dc_pin: 4
    led_pin: 15
    reset_pin: 22
    rotation: 90
    id: tft_ha
    color_palette: GRAYSCALE  # <----- the new config parameter
    lambda: |-
      it.image(0, 0, id(myimage));

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by davet2001** [comment]
I am working on tests now but would value early feedback before I invest much more time in this direction.

**Comment by davet2001** [comment]
Still wanted and working locally without issues.

**Comment by davet2001** [comment]
Still wanted

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Sorry for the delay @davet2001. Could you resolve the conflicts after #2469 and also open a docs PR for the added config?

**Comment by davet2001** [comment]
@jesserockz Conflicts sorted and docs PR added

File: esphome/components/ili9341/display.py [hunk 1]
```diff
@@ -73,3 +81,13 @@ async def to_code(config):
     if CONF_LED_PIN in config:
         led_pin = await cg.gpio_pin_expression(config[CONF_LED_PIN])
         cg.add(var.set_led_pin(led_pin))
+
+    if config[CONF_COLOR_PALETTE] == "GRAYSCALE":
+        cg.add(var.set_buffer_color_mode(ILI9341ColorMode.BITS_8_INDEXED))
+        rhs = []
+        for x in range(256):
+            rhs.extend([HexInt(x), HexInt(x), HexInt(x)])
+        prog_arr = cg.progmem_array(config[CONF_RAW_DATA_ID], rhs)
```
Comment (jesserockz) [inline esphome/components/ili9341/display.py:90]: I feel like a simple greyscale pallete taking up bytes could be avoided, but that can be changed later.

**Review by jesserockz — APPROVED** [review]

File: esphome/components/ili9341/display.py (hunk 1 above)
Comment (davet2001) [reply esphome/components/ili9341/display.py:90]: Yeah, the greyscale is more of a proof of concept, and the simplest case to make a short PR.  The real value is in part 2 #2502 where the colour pallet is added.

**Pull request merged**